source/manual/intro-to-docker-even-more.html.md

```diff
@@ -0,0 +1,398 @@
+
+* Add a new service to our compose file
+
+```
+nginx-proxy:
+  image: jwilder/nginx-proxy:latest
+  ports:
+    - "80:80"
+  volumes:
+    - /var/run/docker.sock:/tmp/docker.sock
+```
+
+The nginx-proxy service will act as a 'front door' for HTTP requests we make in the browser. We give it access to the Docker daemon so it can auto-detect which containers are running and scan their config for which domains they 'own'. It will then proxy HTTP requests to a matching container, on a declared port.
+
+* Change the `app` stack to work with it
+
+```
+depends_on:
+  - postgres
+  - nginx-proxy
+environment:
+  ...
+  VIRTUAL_HOST: content-publisher-demo.intro-to-docker.gov.uk
+expose:
+  - 3000
+```
+
+* Start the app and try it in your browser
+
+```shell
+$mac docker-compose run content-publisher-demo-app
+```
+
+## Step 7: dnsmasq
+
+The reason `content-publisher-demo.intro-to-docker.gov.uk` works in your browser is because we added a custom entry in `/etc/hosts`. This approach doesn't scale to lots of entries, especially if we need to keep them in-sync across a team. We need something more dynamic, but that's not possible with `/etc/hosts`.
+
+As part of [setting-up govuk-docker](https://github.com/alphagov/govuk-docker/blob/master/docs/installation.md) you will have installed [dnsmasq][]. This gives us a little DNS server running on our Mac, which we can use instead of `/etc/hosts` to resolve `*.intro-to-docker.gov.uk` to the localhost.
+
+* Remove the temporary entry in `/etc/hosts`
+
+* Create `/etc/resolver/intro-to-docker.gov.uk`
+
+```
+nameserver 127.0.0.1
+port 53
+```
+
+* Check your Mac is aware of the new rule
+
+```shell
+$mac scutil --dns
+
+...
+
+resolver #9
+  domain   : intro-to-docker.gov.uk
+  nameserver[0] : 127.0.0.1
+  port     : 53
+  flags    : Request A records, Request AAAA records
+  reach    : 0x00030002 (Reachable,Local Address,Directly Reachable Address)
+```
+
+* Edit `/usr/local/etc/dnsmasq.d/development.conf`
+
+```
+address=/dev.gov.uk/127.0.0.1
+address=/intro-to-docker.gov.uk/127.0.0.1
+```
+
+* Restart dnsmasq and your DNS resolver
+
+```shell
+$mac sudo killall -HUP mDNSResponder
+$mac sudo brew services restart dnsmasq
+```
+
+* Start the app and try it in your browser
+
+That's it! You should now have something very similar to the repo we use to develop on GOV.UK. All that remains is for you to go forth, use it, make it work better, and develop some software :-).
+
+## Step 8: Cleanup
+
+During this tutorial you created or changed a lot of files and state on your system. The following indicate how to cleanup your system e.g. if you want to run through the tutorials again.
+
+* Remove all the extra docker state
+
+```shell
+# ignore any errors, you may need to run multiple times
```

Comment:
This instruction is a little confusing. Do I need to re-run these commands until no errors are happening anymore?

Reply:
I've rephrased as a note below the code block - is that better? Re-ordering the commands in the block __should__ mean that you only need to run it once - pretty sure there aren't any cyclic dependencies in there, which was what the old comment was about.

Reply:
That looks better 👍